docs: adding recipe for dynamically importing Images (#5391)
withastro/docs: src/content/docs/en/recipes/dynamically-importing-images.mdx

---
title: Dynamically Import Images
description: Learn how to dynamically import images using Vite's import.meta.glob function
i18nReady: true
type: recipe
---
import FileTree from '~/components/FileTree.astro'

> When working with custom components that include images, there will be times where you will need to dynamically import the image paths. But if the images are located in the `src` directory, it is possible to run into Astro or Vite import errors.  

Reviewer: > But if the images are located in the `src` directory, it is possible to run into Astro or Vite import errors.

You can't (really) import from outside the `src` directory, so this sentence is a bit weird.

In this recipe, you will learn how to dynamically import your images the correct way using Vite's `import.meta.glob` function. 

## Recipe

1. Create a new `assets` folder under the `src` directory and add your images inside that new folder. 

<FileTree>
- src/
  - assets/
    - example-image.jpg
    - example-image-2.jpg
    - example-image-3.jpg
</FileTree>

:::note
`assets` is a popular folder name convention for placing images but you are free to name the folder whatever you like.
::: 

2. Create a new Astro component and import the `<Image>` component.

    ```astro title="src/components/MyCustomCardComponent.astro" 
    ---
     import { Image } from 'astro:assets';
    ---
    ```

3. Create the `props` for your component.

    ```astro title="src/components/MyCustomCardComponent.astro" ins={3}
    ---
     import { Image } from 'astro:assets';

>      const { imagePath, altText, name, age } = Astro.props;

Reviewer 2: ```suggestion
    interface Props {
        imagePath: string;
        altText: string;
        name: string;
        age: number;
    }

     const { imagePath, altText, name, age } = Astro.props;
```

    ---

>     ``` 

Reviewer: Might be useful to show the types here

    ```astro title="src/components/MyCustomCardComponent.astro" ins={4}
    ---
     import { Image } from 'astro:assets';
     const { imagePath, altText, name, age } = Astro.props;
     const images = import.meta.glob('/src/**/*.{jpeg,jpg,png,gif}')
    ---
    ```

5. Create the markup for your component. 

    ```astro title="src/components/MyCustomCardComponent.astro" ins={6-10}
    ---
     import { Image } from 'astro:assets';
     const { imagePath, altText, developer, age } = Astro.props;
     const images = import.meta.glob('/src/**/*.{jpeg,jpg,png,gif}')
    ---
    <div class="card">
        <h2>{developer}</h2>
        <p>Age: {age}</p>
        <Image src={} alt={altText} />
    </div>
    ```

6. Inside the `src` attribute, pass in the `images` object and use bracket notation for the image path. Then make sure to invoke the glob function. 
   
    ```astro title="src/components/MyCustomCardComponent.astro" ins={9}
    ---
     import { Image } from 'astro:assets';
     const { imagePath, altText, developer, age } = Astro.props;

>      const images = import.meta.glob('/src/**/*.{jpeg,jpg,png,gif}')

Reviewer: The return shape of this function is a bit peculiar, it might be helpful to show it so people *understand* how it works

Reviewer 2: I think it's `Record<string, () => Promise<{ default: ImageMetadata }>>`?

---
title: Dynamically Import Images
description: Learn how to dynamically import images using Vite's import.meta.glob function
i18nReady: true
type: recipe
---
import FileTree from '~/components/FileTree.astro'

> When working with [images](/en/guides/images/), there will be times where you will need to dynamically import the image paths. 

Reviewer 3: ```suggestion
Local [images](/en/guides/images/) must be imported into `.astro` files in order to display them. There will be times where you will want or need to dynamically import the image paths of your images instead of explicitly importing each individual image.
```
I like emphasizing that images must be imported here, because I feel like that's something we keep reminding people of!  :smile:  This also gives a bit of context re: what problem this solves: avoiding needing to manually and individually import every image you might want.

> When working with custom components that include images, there will be times where you will need to dynamically import the image paths. 

Reviewer 2: ```suggestion
When working with [images](/en/guides/images/), there will be times where you will need to dynamically import the image paths. 
```

> In this recipe, you will learn how to dynamically import your images the correct way using Vite's `import.meta.glob` function. 

Reviewer 3: ```suggestion
In this recipe, you will learn how to dynamically import your images using Vite's `import.meta.glob` function. You will build a card component that displays the name, age, and photo of a person.
```
Probably don't need to maybe accidentally judge people here! They're just going to learn how to do it via X.  :sweat_smile: 

Also, I like mentioning what we're building up front in this case especially because we're going to define some props right away related to this, but not strictly necessary for importing images to work. I think this gives a little extra helpful context for that!

## Recipe

1. Create a new `assets` folder under the `src` directory and add your images inside that new folder. 

<FileTree>
- src/
  - assets/

>     - example-image.jpg
>     - example-image-2.jpg
>     - example-image-3.jpg

Reviewer 3: ```suggestion
    - avatar-1.jpg
    - avatar-2.png
    - avatar-3.jpeg
```
If these are going to be images associated with people, I think it's a nice touch to name them something like `avatar` or `headshot` or `person` etc. It reinforces the entire narrative around the thing you're building!

Also, since your import below is built to handle different file extensions, it could be an idea to show those off here by making each image a different file type.

</FileTree>

> :::note

Reviewer 3: I **adore** this note. Perfectly written *and* perfectly placed! :purple_heart:

`assets` is a popular folder name convention for placing images but you are free to name the folder whatever you like.
::: 

> 2. Create a new Astro component and import the `<Image>` component.

Reviewer 3: ```suggestion
2. Create a new Astro component for your card and import the `<Image />` component.
```
Just extending the narrative into the instructions. Again, I think tiny bits of context like this help keep the reader focused because things don't feel *too* general.

    ```astro title="src/components/MyCustomCardComponent.astro" 
    ---
     import { Image } from 'astro:assets';
    ---
    ```

> 3. Create the `props` for your component. 

Reviewer 3: ```suggestion
3. Specify the `props` that your component will receive in order to display the necessary information on each card. You can optionally define their types, if you are using TypeScript in your project.
```
I'm thinking something like this is again helpful to reinforce the narrative. (It can be hard to just say, "Here are your props!" but since we've already established what we're trying to build, this is just a gentle reminder, "Hey, we said we were going to display name, age and an image! So, we're setting up the component to be ready to handle that info!"

I love typing the props here! And I think I'd like to call attention to it as optional since I know many of our users aren't using TypeScript. See what you think of the above.

    ```astro title="src/components/MyCustomCardComponent.astro" ins={4-11}
    ---
     import { Image } from 'astro:assets';

     interface Props {
        imagePath: string;
        altText: string;
        name: string;
        age: number;
     }

     const { imagePath, altText, name, age } = Astro.props;
    ---
    ``` 

4. Create a new `images` variable and use the `import.meta.glob` function which returns an object of all of the image paths inside the `assets` folder. 

    ```astro title="src/components/MyCustomCardComponent.astro" ins={12}
    ---
     import { Image } from 'astro:assets';

     interface Props {
        imagePath: string;
        altText: string;
        name: string;
        age: number;
     }
     
     const { imagePath, altText, name, age } = Astro.props;

>      const images = import.meta.glob<Record<string, () => Promise<{ default: ImageMetadata }>>('/src/**/*.{jpeg,jpg,png,gif}')

Reviewer 2: @Princesseuh it's not required to provide the generic right? That's the default signature. Maybe we could just add it in a comment above or below?

Reviewer 3: Agreed with @florian-lefebvre here: I'd love to see that original import function so as not to startle our non-TypeScript users.

Similar to my comment above about defining the prop types as optional, if it's helpful for TS users to know the shape, then after this code snippet there could be a mention separately of it.

    ---
    ```

> 5. Create the markup for your component. 

Reviewer 3: ```suggestion
5. Use the props to create the markup for your card component. 
```
Again, maybe a tiny bit more context to keep the story going?

    ```astro title="src/components/MyCustomCardComponent.astro" ins={14-18}
    ---
     import { Image } from 'astro:assets';

     interface Props {
        imagePath: string;
        altText: string;
        name: string;
        age: number;
     }
     
     const { imagePath, altText, name, age } = Astro.props;
     const images = import.meta.glob<Record<string, () => Promise<{ default: ImageMetadata }>>('/src/**/*.{jpeg,jpg,png,gif}')
    ---
    <div class="card">
        <h2>{developer}</h2>
        <p>Age: {age}</p>
        <Image src={} alt={altText} />
    </div>
    ```

> 6. Inside the `src` attribute, pass in the `images` object and use bracket notation for the image path. Then make sure to invoke the glob function. 

Reviewer 3: ```suggestion
6. Inside the `src` attribute, pass in the `images` object and use bracket notation for the image path. Then make sure to invoke the glob function. 
```
I think I might like a bit of context here explaining that you'll get the `imagePath` as a string as props passed to your component, BUT the cool thing here is that your dynamic import statement is doing the work of the import for you so that you can actually use whatever string is passed! I think this part right here is the star of the show, and I'd love to see these instructions still as wonderfully concise as you've been writing them but helping them connect the dots on this key point.

I know we refrain from a lot of conceptual information in these recipes, and you're providing fabulously explicit instructions which is so important! I do think this is one of those "aha" moments for people that is worth calling out here, since the fact we even need this recipe is an indication that people aren't intuitively picking this up.

Author: I will work on adding more of context here

Author: I was able to add a note 👍

   
    ```astro title="src/components/MyCustomCardComponent.astro" ins={17}
    ---
     import { Image } from 'astro:assets';

     interface Props {
        imagePath: string;
        altText: string;
        name: string;
        age: number;
     }
     
     const { imagePath, altText, name, age } = Astro.props;
     const images = import.meta.glob<Record<string, () => Promise<{ default: ImageMetadata }>>('/src/**/*.{jpeg,jpg,png,gif}')
    ---
    <div class="card">
        <h2>{developer}</h2>
        <p>Age: {age}</p>

>         <Image src={images[imagePath]()} alt={altText} />

Reviewer 3: ```suggestion
         // use the dynamically-imported image specified in props
        <Image src={images[imagePath]()} alt={altText} />
```
Not totally in love with this myself, and I might not have worded it well, but this is a situation where I feel it would be OK to call attention to what's going on with a code comment here. If you feel you can nail the explanation in the text before the code sample and don't need this, then totally fine not to have something here. If you feel you've got a really good concise statement to go here, then I think it could fit, even with a previous explanation. Play around with this entire step 6 as a whole and see what you come up with! This was just an idea that popped into my head.

    </div>
    ```

> 7. Import the component inside an Astro page and pass in the values for the `props`. 

Reviewer 3: ```suggestion
7. Import and use the card component inside an Astro page, passing in the values for the `props`. 
```
I'm a big fan of "import and use" or "import and render" to emphasize that there's going to be two parts to this process.

    ```astro title="src/pages/index.astro" 
    ---
    import MyCustomCardComponent from '../components/MyCustomCardComponent.astro';
    ---
    <MyCustomCardComponent 

>         imagePath="../assets/example-image.jpg"
>         altText="Example image description"
>         developer="John Doe"
>         age={25}

Reviewer 3: ```suggestion
        imagePath="../assets/priya.jpg"
        altText="A headshot of Priya against a brick wall background."
        developer="Priya "
        age={25}
```
So, whatever you decided for naming your images, just make sure it's consistent here. (Those could even be actual names like suggested above. Just as long as they're consistent!)

Also, I'm happy for us to be using more variety of names in the examples we use. I just picked a name at random, but if there's someone you'd like to honour, or a type of name you don't often see recommended, then feel absolutely free to choose anything you want here!

    ```astro title="src/components/MyCustomCardComponent.astro" 
    ---
     import { Image } from 'astro:assets';
    ---
    ```

3. Specify the `props` that your component will receive in order to display the necessary information on each card. You can optionally define their types, if you are using TypeScript in your project.

    ```astro title="src/components/MyCustomCardComponent.astro" ins={4-11}
    ---
     import { Image } from 'astro:assets';

     interface Props {
        imagePath: string;
        altText: string;
        name: string;
        age: number;
     }

     const { imagePath, altText, name, age } = Astro.props;
    ---
    ``` 

4. Create a new `images` variable and use the `import.meta.glob` function which returns an object of all of the image paths inside the `assets` folder. 

    ```astro title="src/components/MyCustomCardComponent.astro" ins={12}
    ---
     import { Image } from 'astro:assets';

     interface Props {
        imagePath: string;
        altText: string;
        name: string;
        age: number;
     }
     
     const { imagePath, altText, name, age } = Astro.props;

>      const images = import.meta.glob('/src/**/*.{jpeg,jpg,png,gif}')

Reviewer 4: I think we can be a little more specific and search the `assets` folder

```suggestion
     const images = import.meta.glob('/src/assets/*.{jpeg,jpg,png,gif}')
```

    ---
    ```

5. Use the props to create the markup for your card component. 

    ```astro title="src/components/MyCustomCardComponent.astro" ins={14-18}
    ---
     import { Image } from 'astro:assets';

     interface Props {
        imagePath: string;
        altText: string;
        name: string;
        age: number;
     }
     
     const { imagePath, altText, name, age } = Astro.props;

>      const images = import.meta.glob('/src/**/*.{jpeg,jpg,png,gif}')

Reviewer 4: Based on other comments, I think we can add a type comment for ts users.
```suggestion
     \\ Type: Record<string, () => Promise<{default: ImageMetadata}>>
     const images = import.meta.glob('/src/**/*.{jpeg,jpg,png,gif}')
```

Reviewer 2: Wrong slashes, you need `//`

Reviewer 3: I'll amend for the commit!  🙌

    ---
    <div class="card">

>         <h2>{developer}</h2>

Reviewer 4: ```suggestion
        <h2>{name}</h2>
```

        <p>Age: {age}</p>
        <Image src={} alt={altText} />
    </div>
    ```

6. Inside the `src` attribute, pass in the `images` object and use bracket notation for the image path. Then make sure to invoke the glob function. 
   
    ```astro title="src/components/MyCustomCardComponent.astro" ins={17}
    ---
     import { Image } from 'astro:assets';

     interface Props {
        imagePath: string;
        altText: string;
        name: string;
        age: number;
     }
     
     const { imagePath, altText, name, age } = Astro.props;
     const images = import.meta.glob('/src/**/*.{jpeg,jpg,png,gif}')
    ---
    <div class="card">
        <h2>{developer}</h2>
        <p>Age: {age}</p>
        <Image src={images[imagePath]()} alt={altText} />
    </div>
    ```

> :::note

Reviewer 2: @sarah11918 wdyt about still putting the type somewhere, maybe at the bottom of the note? It's not the easiest one

Reviewer 2: Looks like Kevin addressed it, feel free to resolve!

`images` is an object that contains all of the image paths inside the `assets` folder. 

```
const images = {
  './assets/avatar-1.jpg': () => import('./assets/avatar-1.jpg'),
  './assets/avatar-2.png': () => import('./assets/avatar-2.png'),
  './assets/avatar-3.jpeg': () => import('./assets/avatar-3.jpeg')
}
```

The `imagePath` prop is a string that contains the path to the image that you want to display. The `import.meta.glob()` is doing the work of finding the image path that matches the `imagePath` prop and handling the import for you.
::: 

7. Import and use the card component inside an Astro page, passing in the values for the `props`. 

    ```astro title="src/pages/index.astro" 
    ---
    import MyCustomCardComponent from '../components/MyCustomCardComponent.astro';
    ---
    <MyCustomCardComponent 

>         imagePath="../assets/priya.jpg"

Reviewer 3: ```suggestion
        imagePath="../assets/avatar-1.jpg"
```
Just a heads up that this should match one of your images mentioned earlier!

>         imagePath="../assets/avatar-1.jpg"

Reviewer 4: I couldn't get the relative path to work on my demo but absolute paths seem to work well.

```suggestion
        imagePath="/src/assets/avatar-1.jpg"
```

Reviewer 3: Curious! I would expect relative paths to work, but this should work too!

---
title: Dynamically Import Images
description: Learn how to dynamically import images using Vite's import.meta.glob function
i18nReady: true
type: recipe
---
import FileTree from '~/components/FileTree.astro'

When working with custom components that include images, there will be times where you will need to dynamically import the image paths. But if the images are located in the `src` directory, it is possible to run into Astro or Vite import errors.  

In this recipe, you will learn how to dynamically import your images the correct way using Vite's `import.meta.glob` function. 

## Recipe

> 1. Create a new `assets` folder under the `src` directory and add your images inside that new folder. 

Reviewer 2: Maybe we can add a notice about the `assets` folder here : it's just a convention but people are free to put images wherever they want. They'll just need to update the string the `glob` function

Author: I added the note 👍

<FileTree>
- src/
  - assets/
    - example-image.jpg
    - example-image-2.jpg
    - example-image-3.jpg
</FileTree>

2. Create a new Astro component and import the `<Image>` component.

    ```astro title="src/components/MyCustomCardComponent.astro" 
    ---
     import { Image } from 'astro:assets';
    ---
    ```

3. Create the `props` for your component.

    ```astro title="src/components/MyCustomCardComponent.astro" ins={3}
    ---
     import { Image } from 'astro:assets';
     const { imagePath, altText, name, age } = Astro.props;
    ---
    ``` 

4. Create a new `images` variable and use the `import.meta.glob` function which returns an object of all of the image paths inside the `assets` folder. 

    ```astro title="src/components/MyCustomCardComponent.astro" ins={4}
    ---
     import { Image } from 'astro:assets';
     const { imagePath, altText, name, age } = Astro.props;

>      const images = import.meta.glob("../assets/*");

Reviewer 2: When using dynamic images in several locations, it can be a bit tedious to write this in several locations with different relatives path (eg `../assets/*` and `../../assets/*`). It could be great to show a more centralized pattern as well as shown on Discord:
```ts
// file located in a more central place
async function imp(filePath) {
  const images = import.meta.glob('/src/**/*.{jpeg,jpg,png,gif}')
  const keys = Object.keys(images)
  const key = keys.find(key => key.includes(filePath))

  if (!key) throw new Error(`Image ${filePath} not found`)

  const image = await images[key]()
  return image.default
}

const a = await imp('/assets/home/hero.jpg')
console.log(a)
```
Also, it may be great to restrict the extension types like in the code snippet above to improve TS typings
